datalake/queue.py

```diff
@@ -0,0 +1,125 @@
+'''manage a queue of datalake files
+
+This allows users to enqueue files to be uploaded to the datalake. An uploader
+process runs that actually does the uploader work.
+
+Under the hood, the queue is a directory which the Uploader watches. The
+Enqueuer enqueues files by setting an extended filesystem attribute with the
+fully-formed metadata for the file to be uploaded, and symlinking it to the
+queue directory. This ensures that the enqueuer fails in the user's face
+instead of silently behind the user's back. The uploader uses inotify to
+monitor the queue directory. When a file arrives, it gets uploaded and the
+symlink deleted.
+
+'''
+from os import environ
+from datalake_common.errors import InsufficientConfiguration
+from xattr import setxattr, getxattr
+import pyinotify
+from logging import getLogger
+import os
+import time
+import shutil
+
+from datalake_common import Metadata
+from datalake import File
+
+
+log = getLogger('datalake-queue')
+
+
+DATALAKE_METADATA_XATTR = 'user.datalake-metadata'
+
+
+class DatalakeQueueBase(object):
+
+    def __init__(self, queue_dir=None):        
+        self.queue_dir = queue_dir or environ.get('DATALAKE_QUEUE_DIR')
+        self._validate_queue_dir()
+
+    def _validate_queue_dir(self):
+        if self.queue_dir is None:
+            raise InsufficientConfiguration('Please specify a queue directory')
+        self.queue_dir = os.path.abspath(self.queue_dir)
+
+
+class Enqueuer(DatalakeQueueBase):
+
+    def enqueue(self, path, **metadata_fields):
+        '''enqueue a file with the specified metadata o be pushed
+
+        Returns the File with complete metadata that will be pushed.
+        '''
+        log.info('Enqueing ' + path)
+        f = File(path, **metadata_fields)
+        setxattr(path, DATALAKE_METADATA_XATTR, f.metadata.json)
+        dest = os.path.join(self.queue_dir, f.metadata['id'])
+        os.symlink(path, dest)        
+        return f
+
+
+class Uploader(DatalakeQueueBase):
+
+    def __init__(self, archive, queue_dir):
+        super(Uploader, self).__init__(queue_dir)
+        self._archive = archive
+
+    class EventHandler(pyinotify.ProcessEvent):
+
+        def __init__(self, callback):
+            super(Uploader.EventHandler, self).__init__()
+            self.callback = callback
+
+        def process_IN_CREATE(self, event):
+            self.callback(event.pathname)
+
+    def _setup_watch_manager(self, timeout):
+        if timeout is not None:
+            timeout = int(timeout * 1000)
+        self._wm = pyinotify.WatchManager()
+        self._handler = Uploader.EventHandler(self._push)
+        self._notifier = pyinotify.Notifier(self._wm, self._handler,
+                                            timeout=timeout)
+        self._wm.add_watch(self.queue_dir, pyinotify.IN_CREATE)
```

Comment:
Good question. We're using symlinks, which get a single IN_CREATE event when they appear. So we're good here. Indeed, if somebody were to copy a file in, the IN_CREATE would cause some trouble. But this directory is internal state of datalake, and the only supported interface is the Enqueuer (which will eventually have a CLI).